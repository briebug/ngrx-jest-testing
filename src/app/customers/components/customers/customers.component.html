<ul>
  <li *ngFor="let customer of customers">Customer: {{customer.id}} - {{customer.name}}
  </li>
</ul>
