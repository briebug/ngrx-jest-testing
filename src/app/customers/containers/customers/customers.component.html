<app-customers [customers]="customers | async"></app-customers>
