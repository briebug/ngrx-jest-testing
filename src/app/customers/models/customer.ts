export interface Customer {
  id: number;
  name: string;
}
