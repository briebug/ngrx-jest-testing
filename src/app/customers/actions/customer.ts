import { Action } from '@ngrx/store';
import { Customer } from '../models/customer';

export enum CustomerActionTypes {
  Load = '[Customer] Load',
  LoadSuccess = '[Customer] Load Success',
  LoadFail = '[Customer] Load Fail',
  Select = '[Customer] Select'
}

export class Load implements Action {
  readonly type = CustomerActionTypes.Load;
}

export class LoadSuccess implements Action {
  readonly type = CustomerActionTypes.LoadSuccess;

  constructor(public payload: Customer[]) {}
}

export class LoadFail implements Action {
  readonly type = CustomerActionTypes.LoadFail;

  constructor(public error: any) {}
}

export class Select implements Action {
  readonly type = CustomerActionTypes.Select;

  constructor(public payload: Customer) {}
}

export type CustomerActions = Load | LoadSuccess | LoadFail | Select;
