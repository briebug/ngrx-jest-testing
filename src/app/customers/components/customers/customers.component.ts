import { Component, OnInit, Input } from '@angular/core';
import { CustomerActions } from '@state/customer/customer.actions';
import { Customer } from '@state/customer/customer.model';

@Component({
  selector: 'app-customers',
  templateUrl: './customers.component.html',
  styleUrls: ['./customers.component.css']
})
export class CustomersComponent implements OnInit {
  @Input() customers: Customer[];

  constructor() { }

  ngOnInit() { }
}
