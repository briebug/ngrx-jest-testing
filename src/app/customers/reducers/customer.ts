import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from '../models/customer';
import { CustomerActions, CustomerActionTypes } from '../actions/customer';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number | null;
  loading: boolean;
  error: any;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      return { ...state, error: { message: 'Error loading customers' } };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
