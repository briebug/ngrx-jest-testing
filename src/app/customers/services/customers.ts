import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs/Observable';
import { map } from 'rxjs/operators';
import { Customer } from '../models/customer';

@Injectable()
export class CustomersService {
  private API_PATH = '/api/customers';

  constructor(private http: HttpClient) {}

  load(): Observable<Customer[]> {
    return this.http.get<Customer[]>(this.API_PATH);
  }
}
