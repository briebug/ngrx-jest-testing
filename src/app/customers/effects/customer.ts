import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from '../actions/customer';

import { Customer } from '../models/customer';
import { map, switchMap, catchError } from 'rxjs/operators';
import { CustomersService } from '../services/customers';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
