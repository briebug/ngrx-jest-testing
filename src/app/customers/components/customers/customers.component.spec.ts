import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { configureTests, ConfigureFn } from '../../../../test-config.helper'

import { Customer } from '@state/customer/customer.model';
import { CustomersComponent } from './customers.component';

describe('CustomersComponent', () => {
  let component: CustomersComponent,
    fixture: ComponentFixture<CustomersComponent>;

  const customers = [
      { id: 1, name: 'test1' },
      { id: 2, name: 'test2' },
      { id: 3, name: 'test3' }
    ];

  beforeEach(
    async(() => {
      const configure: ConfigureFn = testBed => {
        testBed.configureTestingModule({
          declarations: [
            CustomersComponent
          ]
        });
      };

      configureTests(configure).then(testBed => {
        fixture = testBed.createComponent(CustomersComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
      });
    })
  );

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('renders customers markup to snapshot', () => {
    component.customers = customers;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
