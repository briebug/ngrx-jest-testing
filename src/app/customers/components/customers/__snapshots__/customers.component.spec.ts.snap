// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersComponent renders markup to snapshot 1`] = `
<app-customers>
  <p>
    
    customers works!
  
  </p>
  
</app-customers>
`;
